### PR TITLE
Api scheme ten endpoint updates

### PR DESCRIPTION
**Pull request by colinbruce**
#### What
Update the API endpoints to add Scheme Ten specific values
#### Why
Some of the dropdowns now have scheme 10 specific values that were not addressed
#### TODO (wip)
 - [x] write release notes
 - [x] Fee Types
 - [x] Advocate categories

### CODE REVIEW

File: app/models/concerns/roles.rb [hunk 1]
```diff
@@ -8,7 +8,11 @@ module Roles
     klass.validate :roles_valid
 
     klass::ROLES.each do |role|
-      klass.scope role.pluralize.to_sym, -> { matching_role_query([role.to_s]) }
+      if role.split('_').last.numeric?
```
Comment (jsugarman) [inline app/models/concerns/roles.rb]: can leave as is or could also do it by overriding/patching inflections. 

```
ActiveSupport::Inflector.inflections do |inflect|
  inflect.irregular 'agfs_scheme_9', 'agfs_scheme_9'
end
```

not sure if you can do this 'locally' or whether you have to use `config/initializers/inflections.rb`

File: app/models/claim/advocate_claim.rb [hunk 2]
```diff
@@ -180,16 +180,16 @@ def eligible_basic_fee_types
...
+      Fee::BasicFeeType.agfs_scheme_9
     end
 
     def eligible_misc_fee_types
       # TODO: this should return a list based on the current given fee scheme
       # rather than conditionally return scheme 10 specifically
       # TBD once all the fee scheme work is integrated
-      return Fee::MiscFeeType.agfs_scheme_10s if fee_scheme == 'fee_reform'
-      Fee::MiscFeeType.agfs_scheme_9s
+      return Fee::MiscFeeType.agfs_scheme_10 if fee_scheme == 'fee_reform'
```
Comment (jsugarman) [inline app/models/claim/advocate_claim.rb]: just a comment as i do not understand how this effects the api but i rather like the plural version

File: app/interfaces/api/v1/dropdown_data.rb [hunk 3]
```diff
@@ -13,6 +13,10 @@ class DropdownData < API::Helpers::GrapeApiHelper
         def role
           params.role.try(:downcase).try(:to_sym) || :all
         end
+
+        def category
+          params.category.try(:downcase)
```
Comment (jsugarman) [inline app/interfaces/api/v1/dropdown_data.rb]: not a blocker - alternative `params.category&.downcase`

File: app/models/concerns/roles.rb (hunk 1 above)
Comment (lostie) [inline app/models/concerns/roles.rb]: **[not a blocker]** More of a tiny improvement for DRYness.

💭 Consider this instead:

```ruby
scope_name = role.split('_').last.numeric? ? role.to_sym : role.pluralize.to_sym
klass.scope scope_name, -> { matching_role_query([role.to_s]) }
```

or ultimately, if this is a rule that applies to **all strings** ending in numerics:

```ruby
# config/initializers/inflections.rb
ActiveSupport::Inflector.inflections(:en) do |inflect|
  ...
  inflect.plural /^(.*\d)$/, '\1'
end
```

 and in the latter, the code would remain unchanged.

File: app/views/pages/api_release_notes.html.haml [hunk 4]
```diff
@@ -2,6 +2,47 @@
 
 %section.api-documentation
 
+  %hr
+  %h2 19 April 2018
```
Comment (lostie) [inline app/views/pages/api_release_notes.html.haml]: **[not a blocker]** Maybe a 💭 for a possible feature. Seems like if we had some structured release notes (in either DB or even YAML notation) we could more easily updated them, where this template would just recursively render all of that information.

💡Something to thing about...

**Review by jsugarman — APPROVED** [review]
👍 LGTM - just the possible use of inflections for irregular pluralization. as @lostie also suggested.

File: lib/extensions/string_extension.rb [hunk 5]
```diff
@@ -29,6 +29,12 @@ def strftime(format)
     Time.zone.parse(self).strftime(format)
   end
 
+  def numeric?
```
Comment (lostie) [inline lib/extensions/string_extension.rb]: 💭 Consider renaming this method to something that describes more accurately its behaviour (e.g. `has_numeric_ending?`, `ends_with_numeric?`), as the method name, as is, seems misleading.

File: app/models/concerns/roles.rb (hunk 1 above)
Comment (jsugarman) [reply app/models/concerns/roles.rb]: 👍 @lostie - although i think handling the exceptions explicitly via `.irregular` is clearer, a generic numeric inflector might confuse things with patched rails magic

File: app/interfaces/api/v1/dropdown_data.rb [hunk 6]
```diff
@@ -9,9 +9,19 @@ class DropdownData < API::Helpers::GrapeApiHelper
         params :role_filter do
           optional :role, type: String, desc: I18n.t('api.v1.dropdown_data.params.role_filter'), values: %w[agfs lgfs]
         end
+        params :scheme_ten_role_filter do
+          optional :role,
+                   type: String,
+                   desc: I18n.t('api.v1.dropdown_data.params.role_filter'),
+                   values: %w[agfs_scheme_9 agfs_scheme_10 lgfs]
```
Comment (jsugarman) [inline app/interfaces/api/v1/dropdown_data.rb]: As discussed i would include agfs here to and default this to behaving as agfs_scheme_9 - or vice versa - to maintain current functionality -i think this is probably needed as some vendor client apps might use this filtering option to populate list items

File: lib/extensions/string_extension.rb (hunk 5 above)
Comment (colinbruce) [reply lib/extensions/string_extension.rb]: I'm removing it anyway as it's no longer needed but, in truth, it did check the entire string.  I split it in the, also deleted, concern
```ruby
if role.split('_').last.numeric?
```
It was only checking the numeracy of the final string in the array... but anyway, it's going! :)

File: spec/api/v1/dropdown_data_spec.rb [hunk 7]
```diff
@@ -187,6 +187,7 @@ def get_filtered_fee_types(category=nil)
       allow(Settings).to receive(:agfs_reform_advocate_categories).and_return(['QC', 'Leading junior', 'Junior'])
       allow(Settings).to receive(:advocate_categories).and_return(['QC', 'Led junior', 'Leading junior', 'Junior'])
       params.merge!(role: role)
+      binding.pry if role.eql?('agfs_scheme_10s')
```
Comment (lostie) [inline spec/api/v1/dropdown_data_spec.rb]: ✂️

Comment (colinbruce) [reply spec/api/v1/dropdown_data_spec.rb]: 🤦‍♂️ thanks

**Review by lostie — APPROVED** [review]
LGTM 👍

**Pull request merged**